### PR TITLE
Error handling - IN PROGRESS

### PR DESCRIPTION
**Pull request by mphprogrammer**
Handles missing/extra headers when uploading and provides blanket coverage for everything else.

### CODE REVIEW

**Comment by mphprogrammer** [comment]
@patrickvinograd Checked the build log and everything seemed to be OK - what am I missing?

**Comment by patrickvinograd** [comment]
Hard to pick out but right at the bottom there's a file with low coverage:
```
File (/home/jenkins/workspace/bct-data-service_export-bad-PF3C46IFPI3KX4VLWLD4PIGER5SCRBIEAGDBGPHGAS24YCRO6NSQ/app/controllers/concerns/flashable.rb) is only (76.92%) covered. This is below the expected minimum coverage per file of (90.00%).
```

**Comment by patrickvinograd** [comment]
You're now falling through to the rescue block before calling `Flashable#errors_for_alert` - so maybe no longer showing the excerpt of errors from the uploaded file?

File: app/controllers/uploads_controller.rb [hunk 1]
```diff
@@ -9,48 +9,61 @@ def index
...
-    else
-      render :new, alert: errors_for_alert([@upload]), notice: "Failed to upload #{original_filename}."
+      render :new
     end
   end
 
   def show
     @upload = Upload.find_by(id: params[:id])
-    redirect_to uploads_path, alert: ["Upload with id: '#{params[:id]}' not found"] unless @upload
+    return if @upload.present?
```
Comment (saneshark) [inline app/controllers/uploads_controller.rb]: you want to return without rendering?

File: app/controllers/uploads_controller.rb [hunk 2]
```diff
@@ -9,48 +9,61 @@ def index
   end
 
   def new
-    csv_type = params[:csv_type]
-
-    @upload = Upload.new(csv_type: csv_type)
-    @upload.skip_lines = defaults(csv_type)['skip_lines']
-
+    @upload = new_upload(params[:csv_type])
     return if @upload.csv_type.blank? || @upload.csv_type_check?
```
Comment (saneshark) [inline app/controllers/uploads_controller.rb]: I'm confused on this, if you return without rendering how will a user know what happened? You should at least render a flash message or something.  Are you sure you don't mean 'render new and return' ?

File: db/schema.rb [hunk 3]
```diff
@@ -11,7 +11,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20170410221932) do
+ActiveRecord::Schema.define(version: 20170410215512) do
```
Comment (saneshark) [inline db/schema.rb]: not sure why this is necessary if you didn't have any migrations

**Review by saneshark — COMMENTED** [review]
@mphprogrammer can you address these comments and i'll approve?

File: app/controllers/uploads_controller.rb (hunk 1 above)
Comment (mphprogrammer) [inline app/controllers/uploads_controller.rb]: If the upload fails it redirects to the index action, if it succeeds the show view is presented by default.

File: db/schema.rb (hunk 3 above)
Comment (mphprogrammer) [inline db/schema.rb]: There were, although I don't know why - I was presented with a migrations pending error...

Comment (mphprogrammer) [inline db/schema.rb]: I had to fast_forward on the same branch (no changes though!) and then migrate...

File: app/controllers/uploads_controller.rb (hunk 2 above)
Comment (mphprogrammer) [inline app/controllers/uploads_controller.rb]: In light of your changes (removing the generic upload - I should have taken this out.

**Review by saneshark — APPROVED** [review]
I'm satisfied with the input you provided to my questions. If all of @patrickvinograd comments have been addressed too then I think it looks good to merge.

LGTM

**Comment by mphprogrammer** [comment]
@patrickvinograd could you take a peek as well?

**Review by patrickvinograd — APPROVED** [review]
LGTM.

**Pull request merged**